lightning-persister/src/background_processor.rs

```diff
@@ -0,0 +1,219 @@
+use lightning::chain;
+use lightning::chain::chaininterface::{BroadcasterInterface, FeeEstimator};
+use lightning::chain::keysinterface::{ChannelKeys, KeysInterface};
+use lightning::ln::channelmanager;
+use lightning::ln::channelmanager::ChannelManager;
+use lightning::util::logger::Logger;
+use lightning::util::ser::Writeable;
+use std::sync::Arc;
+use std::sync::mpsc::{self, Sender, TryRecvError};
+use std::thread;
+use std::time::{Duration, SystemTime};
+
+/// BackgroundProcesser runs a background thread that monitors whether the ChannelManager it's given on
+/// start has new updates to persist. If it does, then it writes said ChannelManager to disk/backups.
+pub struct BackgroundProcessor {
+	#[cfg(test)]
+	pub thread_terminator: Sender<()>,
+	#[cfg(not(test))]
+	/// Used to send a message to the thread in BackgroundProcessor::start() to exit the thread.
+	thread_terminator: Sender<()>,
+}
+
+impl BackgroundProcessor {
+	/// Start the background thread that checks whether new updates are available for the ChannelManager,
+	/// and writes it to disk if so.
+	/// We also call ChannelManager::timer_chan_freshness_every_min(), since we're already eating the
+	/// cost of running a thread, we may as well pack as many operations into it as make sense.
+	// Mark as must_use because otherwise the result is dropped immediately, resulting in the thread
+	// being terminated.
+	#[must_use]
+	pub fn start<ChanSigner: 'static + ChannelKeys + Writeable, M: 'static + chain::Watch<Keys=ChanSigner>, T: 'static + BroadcasterInterface, K: 'static + KeysInterface<ChanKeySigner=ChanSigner>, F: 'static + FeeEstimator, L: 'static + Logger>(persister: Arc<dyn channelmanager::Persist<ChanSigner, M, T, K, F, L>>, manager: Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>) -> Self {
+		let (sender, receiver) = mpsc::channel();
+		// See ChannelMananger::persistence_lock for more documentation on the persistence process.
+		thread::spawn(move || {
+			let mut current_time = SystemTime::now();
+			loop {
+				let updates_available = manager.block_until_needs_persist(Some(Duration::from_millis(1000)));
```

Comment:
nit: a full second to realize we need to bail out is *really* high. Maybe 100ms.